docs/langref/relay_expr.rst

```diff
@@ -0,0 +1,396 @@
+All function relations must hold at all call sites;
+type checking is thus treated as a constraint-solving problem.
+For more detail on type relations and their implementations,
+please see the documentation on typing in Relay.
+
+See :py:class:`~tvm.relay.expr.Function` for the definition and documentation of function nodes.
+
+Operators
+=========
+
+An operator is a primitive operation not defined in the Relay
+language. Operators are declaredi n the global operator
+registry in C++. Many common operators are backed by TVM's
+Tensor Operator Inventory (`TOPI <https://github.com/dmlc/tvm/tree/master/topi>`__).
+
+but registered in the global environment in either Python
+or C++. Implementations of operators are typically backed by TVM's TOPI.
+
+An operator requires a user to provide an implementation
+of the operator, its type, and any other desired metadata.
+The operator registry is a column-based store where
+operators are keys, so any metadata (which might be referenced
+by optimization passes) may be registered as a new column.
+
+From the perspective of Relay's type system, an operator is a function,
+so operators may be called like any other function and have function
+types. In particular, operator types are registered using a single
+type relation (see the above subsection), typically a relation
+specialized to that operator. For example, the :code:`add` operator
+is registered with the `Broadcast` relation, indicating that the
+arguments of :code:`add` must be tensors and that the return type
+is a tensor whose shape depends on those of its arguments.
+
+Operators are rendered without a sigil (e.g :code:`add`, :code:`subtract`)
+when pretty-printing Relay programs.
+Operators are explicitly contained in the program and are uniquely
+identifiable by pointer.
+
+See :py:class:`~tvm.relay.op.Op` for the definition and documentation
+of operator nodes, demonstrating the infrastructure for registering
+operator metadata. The other files in :py:class:`~tvm.relay.op` give
+handles for generating a call to various pre-registered operators.
+
+Dataflow Fragment
+=================
+
+This subsection covers the set of Relay expressions that do not involve
+control flow. That is, any portion of a program comprised only of these
+expressions corresponds to a pure computation graph without control flow.
+Note that global and local variables are also part of the dataflow fragment.
+
+Constants
+~~~~~~~~~
+
+This node represents a constant tensor value
+(see :py:mod:`~tvm.relay.Value` for more details).
+The constants are represented as :py:class:`~tvm.NDArray`,
+allowing Relay to utilize TVM operators for constant evaluation.
+
+See :py:class:`~tvm.relay.expr.Constant` for its definition and documentation.
+
+Tuple
+~~~~~
+
+The tuple node builds a finite (that is, of statically known size) sequence of heterogeneous data. 
+These tuples match Python's closely. Their fixed length allows for efficient projection of their
```

Comment:
In this sense, is `Tuple` immutable as in Python?

Reply:
All Relay values are immutable, so that is true